core/activity/src/provider/service.rs

```diff
@@ -350,31 +406,41 @@ async fn monitor_activity(
                 break;
             }
 
-            let dt = (Utc::now().timestamp() - usage.timestamp) as f64;
-            if dt > limit_s {
-                log::warn!("activity {} inactive for {}s, destroying", activity_id, dt);
-                enqueue_destroy_evt(
-                    db,
-                    tracker.clone(),
-                    &activity_id,
-                    provider_id,
-                    app_session_id.clone(),
-                )
-                .await;
+            let duration = (Utc::now().timestamp() - usage.timestamp) as f64;
+
+            if duration > unresponsive_limit || duration > inactivity_limit {
+                if state.state.0 != State::Unresponsive {
+                    log::warn!("activity {} unresponsive after {}s", activity_id, duration);
+                    let new_state =
+                        ActivityState::from(StatePair(State::Unresponsive, state.state.1));
+                    prev_state = Some(state);
+                    let _ = tracker
+                        .update_state(activity_id.clone(), State::Unresponsive)
+                        .await;
+                    if let Err(e) = set_persisted_state(&db, &activity_id, new_state).await {
+                        log::error!("cannot update activity {} state: {}", activity_id, e);
+                    }
+                }
 
-                counter!("activity.provider.destroyed.unresponsive", 1);
-                break;
-            } else if state.state.0 != State::Unresponsive && dt >= unresp_s {
-                log::warn!("activity {} unresponsive after {}s", activity_id, dt);
-                let new_state = ActivityState::from(StatePair(State::Unresponsive, state.state.1));
-                prev_state = Some(state);
-                let _ = tracker
-                    .update_state(activity_id.clone(), State::Unresponsive)
+                if duration > inactivity_limit {
+                    log::warn!(
+                        "activity {} inactive for {}s, destroying",
+                        activity_id,
+                        duration
+                    );
+                    enqueue_destroy_evt(
+                        db,
+                        tracker.clone(),
+                        &activity_id,
+                        provider_id,
+                        app_session_id.clone(),
+                    )
                     .await;
-                if let Err(e) = set_persisted_state(&db, &activity_id, new_state).await {
-                    log::error!("cannot update activity {} state: {}", activity_id, e);
+
+                    counter!("activity.provider.destroyed.unresponsive", 1);
+                    break;
                 }
```

Comment:
The only real change is that the `Unresponsive` state will be set after reaching both `unresponsive_limit` or `inactivity_limit`.